### DISCUSSION TITLE
Incremental selection over logical blocks

### DISCUSSION THREAD

**Discussion by matoous** [helix-editor/helix#1328] [Ideas] [answered]
One things I missed in vim when I switched from Goland and other IntelliJ ides was _incremental selection_, I have not seen it done properly by any of the plugins (but maybe I just didn't find the right one). Would this be a feature worth adding to Helix?

**Comment by kirawi** [comment]
I'm not familiar with the keybinds, but if [this](https://www.jetbrains.com/help/resharper/Coding_Assistance__Extend_Shrink_Selection.html) is what you're talking about then I think 100% Helix is all about that. https://docs.helix-editor.com/keymap.html

**Reply by matoous** [reply]
Yes, that's exactly what I am talking about. [Another example](https://res.cloudinary.com/modus-create-inc/image/upload/c_scale,w_607,h_197,dpr_2/f_auto,q_auto:eco,w_auto/f_auto,q_auto:eco,w_auto/f_auto,q_auto:eco/v1610401952/incremental-selection12.gif?_i=AA)

**Reply by kirawi** [reply]
Press `v` to enter extend mode, that's all I recall.

**Reply by matoous** [reply]
I mean: I tried, it is not possible currently afaik. I opened this discussion to discuss the idea of adding it and whether it even makes sense.

**Reply by kirawi** [reply]
You have to press a movement key like `w` or `b` to extend/shrink the selection, or a specialized key like `x`

**Reply by matoous** [reply]
Sorry, maybe I am just using wrong terms or explaining it wrongly. What I mean by _incremental selection_ is selection that works over tree-sitter nodes such as parameters, function calls, function bodies etc. Basically you start at one tree-sitter node and by incrementing the selection you are traversing the tree-sitter tree upwards, towards the root (an vica versa). I updated the discussion title + description to be more specific.

**Reply by kirawi** [reply]
https://docs.helix-editor.com/usage.html#textobjects?

**Comment by sudormrfbin** [comment, marked as answer]
We have `expand_selection` which does exactly what you describe, but it's not bound to a key by default yet because we don't have a matching `shrink_selection` counterpart.

**Reply by matoous** [reply]
Nice! Works pretty well. I wanted to bind it to `C-]` to try it out but with following config it doesn't work for me:
```toml
[keys.normal]
'C' = { ']' = 'expand_selection' }
```
would you please have any idea why?

**Reply by sudormrfbin** [reply]
This is the way to bind `Control+]`:

```toml
[keys.normal]
'C-]' = 'expand_selection'
```
but it's not working for me. A terminal limitation ? Incidentally `A-]` works (`Alt+]`).

**Reply by matoous** [reply]
So I was playing around today trying to create `shrink_selection` and it seems to be quite more complicated than I anticipated.

https://user-images.githubusercontent.com/15747583/147163039-811fb3c7-04a3-47f2-9a33-4c3fe691730b.mov

I don't think there's clean way to do this without persisting the original position of the cursor and shrinking through the syntax three toward node with the cursor. I was also wondering, is it currently possible to have cursor _inside_ selection? I.e. would it be possible to keep the cursor in the original place and expand the selection around it thus removing the need for storing the state separately? (I guess this doesn't fall into the helix design)

**Reply by sudormrfbin** [reply]
It does require some state persistence, which was why it wasn't implemented initially, and it's not possible to put the cursor in between a selection. Can you open a new PR so that there's more visibility ? There were some other ideas put forward some time ago IIRC about how to do this.

**Reply by archseer** [reply]
The idea was to store a stack we push the previous selection onto when expanding, that way shrinking can reverse the expand. If the stack is emptied out, then we select the first child node.

**Reply by archseer** [reply]
For bindings I was thinking `]o`/`[o` since `[` and `]` are used for pairs

**Reply by matoous** [reply]
I gave it a shot: https://github.com/helix-editor/helix/pull/1340, let's move the discussion to the PR.
